### ISSUE TITLE
How to disconnect the field on the new GhostText?

### ISSUE THREAD

**Issue by fregante** [GhostText/GhostText#178]
_Originally posted by @subnut in https://github.com/GhostText/GhostText/issues/172#issuecomment-735098823_:

> If a field is currently connected, then clicking the GhostText extension icon and selecting the connected field should disconnect it.

Yeah the current disconnection logic isn't super clear. In the previous version you would just click the button again and it would disconnect, but now it just looks for more fields.

So the solution I came up with is to right click the button 😬 or to just close the editor tab (which would make more sense, since _you're done with it so you want to go back to the browser_)

Does anyone have any suggestions for an intuitive way to disconnect the editor from the browser?

**Comment by subnut** [comment]
I think that the current method is sufficient... 
but I think that it would be more intuitive if clicking on the extension caused us to enter `Toggle mode` instead of `Register new element` mode. i.e. Clicking on an already-enabled field should disable it, and clicking on a non-enabled field should enable it

**Comment by llinfeng** [comment]
@fregante Ideas to disconnect:

1. How about adding another keyboard shortcut to disconnect? From a couple of months ago, I recall hitting the same (and only) shortcut shall toggle the text fields on and off. For now, hitting the shortcut for "Connect to GhostText" no longer disconnects. The downside is that now we need to remember another key-binding.

2. Personally, I would prefer having an option with the add-on that restores the old behavior. Maybe we can call it "toggle mode"? With one single shortcut, I enter and edit, and hit it again when done. I use GhostText mostly with Jupyter notebooks.

**Comment by fregante** [comment]
> 1. How about adding another keyboard shortcut to disconnect

I like that. Keyboard shortcuts are cheap, as long as I can register an action without setting a default shortcut (so it's up to the user to customize it)

> 2\. I would prefer having an option with the add-on that restores the old behavior

I feel you, the simplicity was good. I wonder if that should be the regular behavior, while still allowing multiple fields by:

- focus the field first, then activating again
- activate via right click: https://github.com/fregante/GhostText/issues/131

**Comment by fregante** [comment]
Makes me wonder if I should go back to the drawing board and exclusively handle **focused** fields:

- toggle the focused field on _activation_

That's it.

You must select the field **first** and **then** you activate GhostText. This should simplify the code too, because there's no more _"Focus (another) field to activate"_ state.

An additional, simple behavior of the button in this case would be:

- if nothing is focused, disconnect every field

**Comment by llinfeng** [comment]
There is definitely a great value in allowing multiple focus+connections and keeping the editor busy syncing up with all the input fields. Along with the gorgeous flow chart, I can see myself rearranging the tabs and panes opened in the editor to make use of the multiple simultaneous connections. Two concerns, though:
1. Practically, how stable are the connections? I managed to crash the Firefox add-on with [vim-ghost](https://github.com/raghur/vim-ghost), where I was clicking on multiple cells in a Jupyterlab notebook: each cell opened a new buffer (with a temp file name) in Vim but the connection quickly stumbled and crashed. With the Chrome extension, I managed to crash the extension by "double-clicking" on the same imput field on this Github page. (I should have pressed the "Connect to Ghost Text" shortcut in between the "double-clicking".
2. More practically, well: for those who can make good use of the multiple focus+connection pairs, do they really need to keep multiple pairs open at the same time? I am thinking in the context of Jupyter notebooks, which is popular among data scientists: if one opts to go for production, instead of editing multiple cells in a `.ipynb` notebook simultaneously with an editor, well, it is advised to write modularized code in `.py` scripts. 

I suppose I can be terribly wrong with the second concern.

----

By the way, is there an **archive** of older versions for the **Firefox** add-on? I see myself doing one-off edits mostly, where I only need one input field to activate. When I am done editing, it helps to disconnect GhostText add-on to keep all things back to normal. By now, I may have talked myself to vote for an extra option for a "toggle mode": when activated, the "Connect to GhostText" shortcut does two things below, 
1. When there is no connection, activate all the candidate fields and wait for a mouseclick
2. When there is an existing connection, disconnect.

If I may recall, is this how GhostText has been working for years?

**Comment by fregante** [comment]
Sorry but most of that is completely off-topic. Please open a new issue

**Comment by fregante** [comment]
Right-clicking is a bit awkward and doesn't work when the icon is in a menu on Safari (it used to be the same in Chrome, but they added hamburgers now)

<img width="320" alt="Screenshot" src="https://user-images.githubusercontent.com/1402241/225007850-7cda1621-180b-4d87-82aa-e68ab7cdab9f.png">

I think GhostText should have a minimal UI to allow disconnection. This would appear sticky on the page:

<img width="221" alt="225006670-b8294791-1344-42e2-bd42-dcdd9bfa923e" src="https://user-images.githubusercontent.com/1402241/225008047-1b3d39e2-9452-4e06-ab3c-ae790296fc87.png">

The only concern is integrating it with the notifications. Maybe:

- no "connected" notification: https://github.com/fregante/GhostText/blob/1b42d6ec4bedf4ed9188066ce3a88b4f9bdc8146/source/ghost-text.js#L126
- add "Switch to your editor" message in the UI (+ "Hide this" button so that it won't appear anymore)

**Comment by Andrew15-5** [comment]
I'm not sure if this is a bug in the new extension, but the disconnect shortcut that I used no longer works. I just updated everything so that the plugin would [work again](https://github.com/fregante/GhostText/issues/317#issuecomment-2336693046). But the only way to disconnect is to close the file in the editor.

Actually, it seems to still work, but at least this time it took about a minute, maybe. I tried disconnecting with no success. Started writing this message and after some time the other tab got activated/focused because the field finally got disconnected.

I can close the file in the editor, but I'd prefer to use the disconnect shortcut in the browser.
